src/Integration.Vsix.UnitTests/Commands/ProjectExcludePropertyToggleCommandTests.cs

```diff
@@ -31,141 +33,79 @@ namespace SonarLint.VisualStudio.Integration.UnitTests.Commands
 
-            // Test case 1: true --toggle--> clears property
-            this.SetExcludeProperty(project, true);
-
-            // Act
-            testSubject.Invoke(command, null);
-
-            // Assert
-            this.VerifyExcludeProperty(project, null);
-
-            // Test case 2: no property --toggle--> true
-            this.SetExcludeProperty(project, null);
-
-            // Act
-            testSubject.Invoke(command, null);
-
-            // Assert
-            this.VerifyExcludeProperty(project, true);
-        }
-
-        [TestMethod]
-        public void ProjectExcludePropertyToggleCommand_Invoke_MultipleProjects_ConsistentPropValues_TogglesValues()
-        {
-            // Arrange
-            OleMenuCommand command = CommandHelper.CreateRandomOleMenuCommand();
-            command.Enabled = true;
-
-            var testSubject = CreateTestSubject();
-
-            var p1 = new ProjectMock("good1.proj");
-            var p2 = new ProjectMock("good2.proj");
-            projectToLanguageMapper.Setup(x => x.HasSupportedLanguage(p1)).Returns(true);
-            projectToLanguageMapper.Setup(x => x.HasSupportedLanguage(p2)).Returns(true);
-
-            this.projectSystem.SelectedProjects = new[] { p1, p2 };
-
-            // Test case 1: all not set --toggle--> all true
-            // Act
-            testSubject.Invoke(command, null);
+            var projectToLanguageMapper = new Mock<IProjectToLanguageMapper>();
+            projectToLanguageMapper.Setup(x => x.HasSupportedLanguage(project)).Returns(true);
 
-            // Assert
-            this.VerifyExcludeProperty(p1, true);
-            this.VerifyExcludeProperty(p2, true);
+            var propertyManager = CreatePropertyManager(project);
+            SetupGetBooleanProperty(propertyManager, project, initialExcludeValue);
 
-            // Test case 2: all true --toggle--> all not set
-            // Arrange
-            this.SetExcludeProperty(p1, true);
-            this.SetExcludeProperty(p2, true);
+            var testSubject = CreateTestSubject(propertyManager.Object, projectToLanguageMapper.Object);
 
             // Act
             testSubject.Invoke(command, null);
 
             // Assert
-            this.VerifyExcludeProperty(p1, null);
-            this.VerifyExcludeProperty(p2, null);
+            VerifyBooleanPropertyCalls(propertyManager, project, expectedSetValue, 1);
         }
 
         [TestMethod]
-        public void ProjectExcludePropertyToggleCommand_Invoke_MultipleProjects_MixedPropValues_SetIsExcludedTrue()
+        [DataRow(true, true, null, null, 2)]
+        [DataRow(false, false, true, true, 2)]
+        [DataRow(null, false, true, true, 1)]
+        [DataRow(null, true, true, true, 1)]
+        public void Invoke_MultipleProjects_TogglesValues(bool? initialExcludeValue1, bool? initialExcludeValue2,
+            bool? expectedSetValue1, bool? expectedSetValue2, int nrOfCallsToGetBoolean)
```

Comment:
Here I merge two tests into one, 1. MultipleProjects same properties 2. Multiple Projects Different properties.

The only thing I don't really like is the nrOfCallsToGetBoolean but my brain is fried. If you have a better suggesting lmk!

Reply:
My brain hurt trying to work out what the product code was supposed to be doing...

Perhaps the test is being over-specific i.e. it is testing too much. The most important bit of the functionality here is whether the properties are set correctly, not how many times they are read.
Also, it looks like all of the properties should be set to the same value (either `null` or `true`), so `expectedSetValue2` shouldn't be necessary.


If I understand the intention of the code correctly, it's the following:
* if the initial property values are inconsistent then all properties should be set to `true`.
* if the initial property values are identical then they should be toggled: (`true` -> `null`, `false`/`null` -> true)


Assuming all of the above is correct, the test could be simplified to three input values e.g.

```
        [TestMethod]
        [DataRow(true, true, null)]   // Same across projects -> toggle
        [DataRow(false, false, true)] // Same across projects -> toggle
        [DataRow(null, null, true)]   // Same across projects (null === false) -> set to true
        [DataRow(null, false, true)]  // Same across projects (null === false) -> set to true
        [DataRow(null, true, true)]   // Different across projects -> set to true
        [DataRow(false, true, true)]  // Different across projects -> set to true
        public void Invoke_MultipleProjects_ExpectedValueIsSet(bool? initialExcludeValue1, bool? initialExcludeValue2,
            bool? expectedSetValue)
        {
           ...
        }
```